File: src/test/java/cube8540/oauth/authentication/users/application/DefaultUserManagementServiceTest.java
package cube8540.oauth.authentication.users.application;

import cube8540.oauth.authentication.users.domain.User;
import cube8540.oauth.authentication.users.domain.UserCredentialsKeyGenerator;
import cube8540.oauth.authentication.users.domain.UserErrorCodes;
import cube8540.oauth.authentication.users.domain.UserNotFoundException;
import cube8540.oauth.authentication.users.domain.UserRegisterException;
import cube8540.oauth.authentication.users.domain.UserRepository;
import cube8540.oauth.authentication.users.domain.UserValidatorFactory;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.security.crypto.password.PasswordEncoder;

import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.ENCODED_PASSWORD;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.PASSWORD;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.RAW_CREDENTIALS_KEY;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.RAW_USERNAME;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.USERNAME;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeCountingUserRepository;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeDefaultUser;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeEmptyUserRepository;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeErrorValidatorFactory;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeKeyGenerator;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makePasswordEncoder;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeUserRegisterRequest;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeUserRepository;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeValidatorFactory;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@DisplayName("기본 유저 계정 관리 서비스 테스트")
class DefaultUserManagementServiceTest {

    @Test
    @DisplayName("유저 카운팅")
    void userCounting() {
        long randomCount = (long) (Math.random() * 100);
        UserRepository repository = makeCountingUserRepository(USERNAME, randomCount);
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCODED_PASSWORD);
        DefaultUserManagementService service = new DefaultUserManagementService(repository, encoder);

        long result = service.countUser(RAW_USERNAME);
        assertEquals(randomCount, result);
    }

    @Test
    @DisplayName("저장소에 없는 유저 프로필 검색")
    void loadUserProfileToNotRegisteredInRepository() {
        UserRepository repository = makeEmptyUserRepository();
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCODED_PASSWORD);

        DefaultUserManagementService service = new DefaultUserManagementService(repository, encoder);

        assertThrows(UserNotFoundException.class, () -> service.loadUserProfile(RAW_USERNAME));
    }

    @Test
    @DisplayName("이미 저장소에 등록된 유저 영속화")
    void persistUserToAlreadyRegisteredInRepository() {
        User user = makeDefaultUser();
        UserRegisterRequest request = makeUserRegisterRequest();
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCODED_PASSWORD);
        UserRepository repository = makeUserRepository(USERNAME, user);

        DefaultUserManagementService service = new DefaultUserManagementService(repository, encoder);

        UserRegisterException e = assertThrows(UserRegisterException.class, () -> service.registerUser(request));
        assertEquals(UserErrorCodes.EXISTS_IDENTIFIER, e.getCode());
    }

    @Test
    @DisplayName("유효 하지 않은 데이터 유저 등록")
    void persistUserWithNotAllowedData() {
        UserRepository repository = makeEmptyUserRepository();
        UserValidatorFactory factory = makeErrorValidatorFactory(new TestUserException());
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCODED_PASSWORD);
        UserRegisterRequest registerRequest = makeUserRegisterRequest();

        DefaultUserManagementService service = new DefaultUserManagementService(repository, encoder);
        service.setValidatorFactory(factory);

        assertThrows(TestUserException.class, () -> service.registerUser(registerRequest));
    }

    @Test
    @DisplayName("저장소에 유저 등록")
    void persistUserInRepository() {
        UserRepository repository = makeEmptyUserRepository();
        UserValidatorFactory factory = makeValidatorFactory();
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCODED_PASSWORD);
        UserCredentialsKeyGenerator keyGenerator = makeKeyGenerator(RAW_CREDENTIALS_KEY);
        UserRegisterRequest registerRequest = makeUserRegisterRequest();

        DefaultUserManagementService service = new DefaultUserManagementService(repository, encoder);
        ArgumentCaptor<User> userCaptor = ArgumentCaptor.forClass(User.class);
        service.setValidatorFactory(factory);
        service.setKeyGenerator(keyGenerator);

        service.registerUser(registerRequest);
        verify(repository, times(1)).save(userCaptor.capture());
        assertEquals(USERNAME, userCaptor.getValue().getUsername());
        assertEquals(ENCODED_PASSWORD, userCaptor.getValue().getPassword());
        assertEquals(RAW_CREDENTIALS_KEY, userCaptor.getValue().getCredentialsKey().getKeyValue());
    }

    @Test
    @DisplayName("저장소에 저장 되지 않은 유저 삭제")
    void removeUserToNotRegisteredInRepository() {
        UserRepository repository = makeEmptyUserRepository();
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCODED_PASSWORD);

        DefaultUserManagementService service = new DefaultUserManagementService(repository, encoder);
        assertThrows(UserNotFoundException.class, () -> service.removeUser(RAW_USERNAME));
    }

    @Test
    @DisplayName("저장소에서 유저 삭제")
    void removeUser() {
        User user = makeDefaultUser();
        UserRepository repository = makeUserRepository(USERNAME, user);
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCODED_PASSWORD);
        ArgumentCaptor<User> userCaptor = ArgumentCaptor.forClass(User.class);

        DefaultUserManagementService service = new DefaultUserManagementService(repository, encoder);

        service.removeUser(RAW_USERNAME);
        verify(repository, times(1)).delete(userCaptor.capture());
        assertEquals(user, userCaptor.getValue());
    }

    private static class TestUserException extends RuntimeException {}
}

File: src/test/kotlin/cube8540/oauth/authentication/user/application/DefaultUserManagementServiceTest.kt
package cube8540.oauth.authentication.user.application

import cube8540.oauth.authentication.UnitTestValidationException
import cube8540.oauth.authentication.error.message.ErrorCodes
import cube8540.oauth.authentication.users.application.DefaultUserManagementService
import cube8540.oauth.authentication.users.application.UserRegisterRequest
import cube8540.oauth.authentication.users.domain.Uid
import cube8540.oauth.authentication.users.domain.User
import cube8540.oauth.authentication.users.domain.UserCredentialsKey
import cube8540.oauth.authentication.users.domain.UserCredentialsKeyGenerator
import cube8540.oauth.authentication.users.domain.UserNotFoundException
import cube8540.oauth.authentication.users.domain.UserRegisterException
import cube8540.oauth.authentication.users.domain.UserRepository
import cube8540.oauth.authentication.users.domain.UserUidGenerator
import cube8540.oauth.authentication.users.domain.UserValidatorFactory
import cube8540.oauth.authentication.users.domain.Username
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import java.util.Optional
import kotlin.random.Random
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.security.crypto.password.PasswordEncoder

class DefaultUserManagementServiceTest {
    private val repository: UserRepository = mockk(relaxed = true)
    private val passwordEncoder: PasswordEncoder = mockk()
    private val keyGenerator: UserCredentialsKeyGenerator = mockk()
    private val uidGenerator: UserUidGenerator = mockk()
    private val validatorFactory: UserValidatorFactory = mockk()

    private val service = DefaultUserManagementService(repository, passwordEncoder)

    init {
        service.keyGenerator = keyGenerator
        service.uidGenerator = uidGenerator
        service.validatorFactory = validatorFactory

        every { repository.save(any()) } returnsArgument 0
    }

    @Nested
    inner class CountingTest {

        @Test
        fun `user counting`() {
            val count = Random.nextLong(0, 100)

            every { repository.countByUsername(Username("username")) } returns count

            val results = service.countUser("username")
            assertThat(results).isEqualTo(count)
        }
    }

    @Nested
    inner class LoadProfileTest {

        @Test
        fun `request user is not registered in repository`() {
            every { repository.findById(Username("username")) } returns Optional.empty()

            val thrown = catchThrowable { service.loadUserProfile("username") }
            assertThat(thrown).isInstanceOf(UserNotFoundException::class.java)
        }

        @Test
        fun `request user is registered in repository`() {
            val user: User = mockk(relaxed = true) {
                every { uid } returns Uid("uid")
                every { username } returns Username("username")
            }

            every { repository.findById(Username("username")) } returns Optional.of(user)

            val result = service.loadUserProfile("username")
            assertThat(result.uid).isEqualTo("uid")
            assertThat(result.username).isEqualTo("username")
        }
    }

    @Nested
    inner class RegisterNewUserTest {

        @Test
        fun `user is already registered in repository`() {
            val registerRequest = UserRegisterRequest("username", "password")

            every { repository.countByUsername(Username("username")) } returns 1

            val thrown = catchThrowable { service.registerUser(registerRequest) }
            assertThat(thrown).isInstanceOf(UserRegisterException::class.java)
            assertThat(((thrown) as UserRegisterException).code)
                .isEqualTo(ErrorCodes.EXISTS_IDENTIFIER)
        }

        @Test
        fun `request user data is invalid`() {
            val registerRequest = UserRegisterRequest("username", "password")

            every { uidGenerator.generateUid() } returns Uid("uid")
            every { repository.countByUsername(Username("username")) } returns 0
            every { validatorFactory.createValidator(any()) } returns mockk {
                every { result } returns mockk {
                    every { hasErrorThrows(any()) } throws UnitTestValidationException()
                }
            }

            val thrown = catchThrowable { service.registerUser(registerRequest) }
            assertThat(thrown).isInstanceOf(UnitTestValidationException::class.java)
        }

        @Test
        fun `register successful`() {
            val userCaptor = slot<User>()
            val registerRequest = UserRegisterRequest("username", "password")

            every { uidGenerator.generateUid() } returns Uid("uid")
            every { keyGenerator.generateKey() } returns UserCredentialsKey("key")
            every { passwordEncoder.encode("password") } returns "encodedPassword"
            every { repository.countByUsername(Username("username")) } returns 0
            every { repository.save(capture(userCaptor)) } returnsArgument 0
            every { validatorFactory.createValidator(any()) } returns mockk {
                every { result } returns mockk {
                    every { hasErrorThrows(any()) } just Runs
                }
            }

            service.registerUser(registerRequest)
            assertThat(userCaptor.isCaptured).isTrue
            assertThat(userCaptor.captured.uid).isEqualTo(Uid("uid"))
            assertThat(userCaptor.captured.username).isEqualTo(Username("username"))
            assertThat(userCaptor.captured.password).isEqualTo("encodedPassword")
            assertThat(userCaptor.captured.credentialsKey?.keyValue).isEqualTo("key")
        }
    }

    @Nested
    inner class RemoveUserTest {

        @Test
        fun `request user is not registered in repository`() {
            every { repository.findById(Username("username")) } returns Optional.empty()

            val thrown = catchThrowable { service.removeUser("username") }
            assertThat(thrown).isInstanceOf(UserNotFoundException::class.java)
        }

        @Test
        fun `request user is registered in repository`() {
            val user: User = mockk(relaxed = true)

            every { repository.findById(Username("username")) } returns Optional.of(user)

            service.removeUser("username")
            verify { repository.delete(user) }
        }
    }
}